ciao-cli/external-ips.go

```diff
@@ -0,0 +1,870 @@
+	}
+
+	fmt.Printf("Requested external IP for: %s\n", cmd.instanceID)
+
+	return nil
+}
+
+type externalIPListCommand struct {
+	Flag     flag.FlagSet
+	template string
+}
+
+func (cmd *externalIPListCommand) usage(...string) {
+	fmt.Fprintf(os.Stderr, `usage: ciao-cli [options] external-ip list [flags]
+
+List all mapped external IPs.
+
+The list flags are:
+
+`)
+	cmd.Flag.PrintDefaults()
+
+	fmt.Fprintf(os.Stderr, `
+The template passed to the -f option operates on a
+
+[]%s
+`, externalIPTemplate)
+	fmt.Fprintln(os.Stderr, templateFunctionHelp)
+	os.Exit(2)
+}
+
+func (cmd *externalIPListCommand) parseArgs(args []string) []string {
+	cmd.Flag.StringVar(&cmd.template, "f", "", "Template used to format output")
+	cmd.Flag.Usage = func() { cmd.usage() }
+	cmd.Flag.Parse(args)
+	return cmd.Flag.Args()
+}
+
+func (cmd *externalIPListCommand) run(args []string) error {
+	var IPs []types.MappedIP
+
+	url, ver, err := getCiaoExternalIPsResource()
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	resp, err := sendCiaoRequest("GET", url, nil, nil, &ver)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		fatalf("External IP list failed: %s", resp.Status)
+	}
+
+	err = unmarshalHTTPResponse(resp, &IPs)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	if cmd.template != "" {
+		return outputToTemplate("external-ip-list", cmd.template,
+			&IPs)
+	}
+
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 0, 1, 1, ' ', 0)
+	fmt.Fprintf(w, "#\tExternalIP\tInternalIP")
+	if checkPrivilege() {
+		fmt.Fprintf(w, "\tInstanceID\tTenantID\tPoolName\n")
+	}
+
+	for i, IP := range IPs {
+		fmt.Fprintf(w, "%d", i+1)
+		fmt.Fprintf(w, "\t%s", IP.ExternalIP)
+		fmt.Fprintf(w, "\t%s", IP.InternalIP)
+		if IP.InstanceID != "" {
+			fmt.Fprintf(w, "\t%s", IP.InstanceID)
+		}
+
+		if IP.TenantID != "" {
+			fmt.Fprintf(w, "\t%s", IP.TenantID)
+		}
+
+		if IP.PoolName != "" {
+			fmt.Fprintf(w, "\t%s", IP.PoolName)
+		}
+
+		fmt.Fprintf(w, "\n")
+
+		w.Flush()
+	}
+	return nil
+}
+
+type externalIPUnMapCommand struct {
+	address string
+	Flag    flag.FlagSet
+}
+
+func (cmd *externalIPUnMapCommand) usage(...string) {
+	fmt.Fprintf(os.Stderr, `usage: ciao-cli [options] external-ip unmap [flags]
+
+Unmap a given external IP.
+
+The unmap flags are:
+
+`)
+	cmd.Flag.PrintDefaults()
+	os.Exit(2)
+}
+
+func (cmd *externalIPUnMapCommand) parseArgs(args []string) []string {
+	cmd.Flag.StringVar(&cmd.address, "address", "", "External IP to unmap.")
+	cmd.Flag.Usage = func() { cmd.usage() }
+	cmd.Flag.Parse(args)
+	return cmd.Flag.Args()
+}
+
+func (cmd *externalIPUnMapCommand) run(args []string) error {
+	if cmd.address == "" {
+		errorf("Missing required -instance parameter")
+		cmd.usage()
+	}
+
+	url, err := getExternalIPRef(cmd.address)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	ver := api.ExternalIPsV1
+
+	resp, err := sendCiaoRequest("DELETE", url, nil, nil, &ver)
+	if err != nil {
+		fatalf(err.Error())
+	}
+
+	defer resp.Body.Close()
```

Comment:
Yes, sorry I got confused.  I thought the code was closing the body passed into the sendCiaoRequest function.  But it's not.  It's closing the body owned by the response.